apps/compute-cf.yml.j2

```diff
@@ -63,7 +73,70 @@ Resources:
 
             --==BOUNDARY==--
 
-  ComputeEnvironment:
+  SharedComputeEnvironment:
+      Type: AWS::Batch::ComputeEnvironment
+      Properties:
+        ServiceRole: !GetAtt BatchServiceRole.Arn
+        Type: MANAGED
+        ComputeResources:
+          Type: SPOT
+          AllocationStrategy: SPOT_PRICE_CAPACITY_OPTIMIZED
+          MinvCpus: 0
+          MaxvCpus: !Ref MaxvCpus
+          InstanceTypes: !Ref InstanceTypes
+          ImageId: !Ref AmiId
+          Subnets: !Ref SubnetIds
+          InstanceRole: !Ref InstanceProfile
+          SecurityGroupIds:
+            - !Ref SecurityGroup
+          LaunchTemplate:
+            LaunchTemplateId: !Ref SharedLaunchTemplate
+            Version: !GetAtt SharedLaunchTemplate.LatestVersionNumber
+          Tags:
+            Name: !Ref AWS::StackName
+
+  SharedJobQueue:
+    Type: AWS::Batch::JobQueue
+    Properties:
+      Priority: 1
+      ComputeEnvironmentOrder:
+        - ComputeEnvironment: !Ref SharedComputeEnvironment
+          Order: 1
+      SchedulingPolicyArn: !Ref SchedulingPolicy
+
+  SchedulingPolicy:
+    Type: AWS::Batch::SchedulingPolicy
+
+  {% for job_type, job_spec in job_types.items() if 'Shared' not in job_spec['compute_environment']['name'] %}
+  {% set env = job_spec['compute_environment'] %}
+  {% set name = env['name'] %}
+  {% set instance_types = env['instance_types'] if 'instance_types' in env else ['!Ref InstanceTypes'] %}
+  {% set ami_id = env['ami_id'] if 'ami_id' in env else '!Ref AmiId' %}
```

Comment:
You might be able to use the `dict.get` method here, not sure if Jinja supports it.